---
title: "Class19: Pertussis and the CMI-PB project"
author: "Wanning Cui"
format: pdf
---
```{r}
library(tidyverse)
```

```{r}
#|echo:FALSE
cdc <- data.frame(year = c(1922L,1923L,1924L,1925L,
                                          1926L,1927L,1928L,1929L,1930L,1931L,
                                          1932L,1933L,1934L,1935L,1936L,
                                          1937L,1938L,1939L,1940L,1941L,1942L,
                                          1943L,1944L,1945L,1946L,1947L,
                                          1948L,1949L,1950L,1951L,1952L,
                                          1953L,1954L,1955L,1956L,1957L,1958L,
                                          1959L,1960L,1961L,1962L,1963L,
                                          1964L,1965L,1966L,1967L,1968L,1969L,
                                          1970L,1971L,1972L,1973L,1974L,
                                          1975L,1976L,1977L,1978L,1979L,1980L,
                                          1981L,1982L,1983L,1984L,1985L,
                                          1986L,1987L,1988L,1989L,1990L,
                                          1991L,1992L,1993L,1994L,1995L,1996L,
                                          1997L,1998L,1999L,2000L,2001L,
                                          2002L,2003L,2004L,2005L,2006L,2007L,
                                          2008L,2009L,2010L,2011L,2012L,
                                          2013L,2014L,2015L,2016L,2017L,2018L,
                                          2019L,2020L,2021L),
         cases = c(107473,164191,165418,152003,
                                          202210,181411,161799,197371,
                                          166914,172559,215343,179135,265269,
                                          180518,147237,214652,227319,103188,
                                          183866,222202,191383,191890,109873,
                                          133792,109860,156517,74715,69479,
                                          120718,68687,45030,37129,60886,
                                          62786,31732,28295,32148,40005,
                                          14809,11468,17749,17135,13005,6799,
                                          7717,9718,4810,3285,4249,3036,
                                          3287,1759,2402,1738,1010,2177,2063,
                                          1623,1730,1248,1895,2463,2276,
                                          3589,4195,2823,3450,4157,4570,
                                          2719,4083,6586,4617,5137,7796,6564,
                                          7405,7298,7867,7580,9771,11647,
                                          25827,25616,15632,10454,13278,
                                          16858,27550,18719,48277,28639,32971,
                                          20762,17972,18975,15609,18617,
                                          6124,2116)
       )
  
```

```{r}
ggplot(cdc) +
  aes(x=year, y=cases) +
  geom_point() +
  geom_line()
```
>Q2.

```{r}
ggplot(cdc) +
  aes(x=year, y=cases) +
  geom_point() +
  geom_line() +
  geom_vline(xintercept=1946,color="red")+
  geom_vline(xintercept=1996,color="blue")+
  geom_vline(xintercept=2019,color="darkgreen")
```
>Q3. 

The pertussis cases increase again after the introduction of the vaccine (e.g. around year 2012). This could possibly be due to protests against vaccination or development of more sensitive antibodies.


#CMI-PB project

The CMI-PB project collects and makes available data on the immune response to Pertussis booster vaccination.

We will access this data via the API.
We will use the **jsonlite** package to access the data using the `read_json()` function.

```{r}
library(jsonlite)
```

```{r}
subject <- read_json("https://www.cmi-pb.org/api/subject", simplifyVector = TRUE)
```

```{r}
head(subject)
```

>Q4. How many aP and wP infancy vaccinated subjects are in the dataset?

```{r}
table(subject$infancy_vac)
```

>Q5. How many Male and Female subjects/patients are in the dataset?

```{r}
table(subject$biological_sex)
```

>Q6. What is the breakdown of race and biological sex?

```{r}
table(subject$biological_sex,subject$race)
```

>Q. Make a histogram of the subject age distribution and facet by infancy_vac

```{r}
library(lubridate)
```

```{r}
today() - mdy("06-11-2000")
```

```{r}
time_length( today() - mdy("06-11-2000"),  "years")
```

>Q7.

```{r}
subject$age <- ymd(subject$date_of_boost) - ymd(subject$year_of_birth)
```

```{r}
subject$age_years <- time_length(subject$age,"years")
```

```{r}
ggplot(subject)+
  aes(age_years,
      fill=infancy_vac)+
  facet_wrap(vars(infancy_vac),ncol=1)+
  geom_histogram()
```
>Q8.Determine the age of all individuals at time of boost?

```{r}
int <- ymd(subject$date_of_boost) - ymd(subject$year_of_birth)
age_at_boost <- time_length(int, "year")
head(age_at_boost)
```

>Q9.With the help of a faceted boxplot or histogram (see below), do you think these two groups are significantly different?

Yes.

There are 3 main datasets in the CMI-PB project at the time of writing.

```{r}
table(subject$dataset)
```

```{r}
specimen <- read_json("https://www.cmi-pb.org/api/specimen",simplifyVector=TRUE)
titer <- read_json("https://www.cmi-pb.org/api/v4/plasma_ab_titer",simplifyVector=TRUE)
```

```{r}
head(specimen)
```
```{r}
head(titer)
```

>Q9. I want to (join) merge the specimen and subject tables together.

```{r}
meta <- inner_join(specimen, subject)
head(meta)
```

>Q10.

```{r}
abdata <- inner_join(titer, meta)
dim(abdata)
```

>Q11. How many specimens (i.e. entries in abdata) do we have for each isotype?

```{r}
table(abdata$isotype)
```

>Q12. What are the different $dataset values in abdata and what do you notice about the number of rows for the most “recent” dataset?

```{r}
table(abdata$dataset)
```
The number of rows are decreasing.

>Q13. 

```{r}
igg <- abdata %>% filter(isotype == "IgG")
head(igg)
```
```{r}
ggplot(igg)+
  aes(MFI_normalised,antigen)+
  geom_boxplot()+
  xlim(0,75)+
  facet_wrap(vars(visit),nrow=2)
```

```{r}
oops <- abdata %>% filter(antigen == "Fim2/3")
table(oops$dataset)
```

>Q14. What antigens show differences in the level of IgG antibody titers recognizing them over time? Why these and not others?

The antigens are Fim2/3 and PT. This is because they are immunogenic.

>Q15. 

```{r}
filter(igg, antigen=="OVA") %>%
  ggplot() +
  aes(MFI_normalised, col=infancy_vac) +
  geom_boxplot(show.legend = TRUE) +
  facet_wrap(vars(visit)) +
  theme_bw()
```

```{r}
filter(igg, antigen=="FIM2/3") %>%
  ggplot() +
  aes(MFI_normalised, col=infancy_vac) +
  geom_boxplot(show.legend = TRUE) +
  facet_wrap(vars(visit)) +
  theme_bw()
```

>Q16. What do you notice about these two antigens time courses and the PT data in particular?

PT levels clearly rise over time and far exceed those of OVA. They also appear to peak at visit 5 and then decline. This trend appears similar for wP and aP subjects.

>Q17.Do you see any clear difference in aP vs. wP responses?

The aP responses for OVA antigen is a bit higher than wP responses. However, the wP responses for PT antigen is higher than aP responses.


Select (or filter) for the 2021 dataset and isotype IgG
I want a time course (`planned_day_relative_to_boost`) of IgG versus (`MFI_normalised`)for"PT"angigen

```{r}
igpt.21 <- abdata %>% filter (dataset == "2021_dataset",
                            isotype == "IgG",
                            antigen == "PT")
```

```{r}
ggplot(igpt.21)+
  aes(planned_day_relative_to_boost,
      MFI_normalised,
      col=infancy_vac)+
  geom_point()+
  geom_line(aes(group=subject_id), linewidth=0.5,alpha=0.5)+
  geom_smooth(se=FALSE,span=0.4,linewidth=3)
```

```{r}
igpt.21 <- abdata %>% filter (dataset == "2020_dataset",
                            isotype == "IgG",
                            antigen == "PT")
ggplot(igpt.21)+
  aes(planned_day_relative_to_boost,
      MFI_normalised,
      col=infancy_vac)+
  geom_point()+
  geom_line(aes(group=subject_id), linewidth=0.5,alpha=0.5)+
  geom_smooth(se=FALSE,span=0.4,linewidth=3)
```

>Q18. Does this trend look similar for the 2020 dataset?

Yes.

```{r}
igpt.22 <- abdata %>% filter (dataset == "2022_dataset",
                            isotype == "IgG",
                            antigen == "PT")
ggplot(igpt.22)+
  aes(planned_day_relative_to_boost,
      MFI_normalised,
      col=infancy_vac)+
  geom_point()+
  geom_line(aes(group=subject_id))+
  geom_vline(xintercept=0) +
  geom_vline(xintercept=14)

```



Make a plot of IgG versus MFI_normalised for "PT" antigen and facet by dataset and infancy_vac

```{r}
ggplot(igpt.21)+
  aes(planned_day_relative_to_boost,
      MFI_normalised,
      col=infancy_vac)+
  geom_point()+
  geom_line(aes(group=subject_id), linewidth=0.5,alpha=0.5)+
  geom_smooth(se=FALSE,span=0.4,linewidth=3)+
  facet_wrap(vars(dataset,infancy_vac),ncol=1)
```

```{r}
url <- "https://www.cmi-pb.org/api/v2/rnaseq?versioned_ensembl_gene_id=eq.ENSG00000211896.7"
rna <- read_json(url, simplifyVector = TRUE) 
```

```{r}
ssrna <- inner_join(rna, meta)
```

>Q19. 

```{r}
ggplot(ssrna) +
  aes(x=visit, y=tpm, group=subject_id) +
  geom_point() +
  geom_line(alpha=0.2)
```
>Q20.: What do you notice about the expression of this gene (i.e. when is it at it’s maximum level)?

The gene expression gradually increases and reaches peak at 4th visit.

>Q21. Does this pattern in time match the trend of antibody titer data? If not, why not?

No. The antibody titer data shows at later timepoints. This is because antibodies made by cells live longer.


```{r}
ggplot(ssrna) +
  aes(tpm, col=infancy_vac) +
  geom_boxplot() +
  facet_wrap(vars(visit))
```

```{r}
ssrna %>%  
  filter(visit==4) %>% 
  ggplot() +
    aes(tpm, col=infancy_vac) + geom_density() + 
    geom_rug()
```

